File: pyhessref/hess_trait_unrestricted.py
import numpy as np
from abc import ABC, abstractmethod


class UHessCoreAPI(ABC):
    """Abstract class for Hessian-related API for unrestricted SCF core components.

    Term Explanation
    ----------------

    See `RHessCoreAPI` (restricted version). The interpretation of *core* is identical.

    UHF Conventions
    ---------------

    For UHF, the molecular orbital descriptors are stored per spin:
    - ``mo_coeff``: shape ``[2, nao, nmo]``
    - ``mo_occ``:   shape ``[2, nmo]`` (occupied orbitals have occupation 1, not 2)
    - ``mo_energy``: shape ``[2, nmo]``

    Since core/skeleton hessian terms depend only on the *total* density matrix
    ``D = D^alpha + D^beta`` and the AO-basis derivative is spin-independent,
    the contract here looks identical to the restricted case from the caller's
    point of view -- the implementing class is responsible for assembling the total
    density internally if needed.
    """

    @abstractmethod
    def make_skeleton_hess(
        self,
        mo_coeff: np.ndarray,
        mo_occ: np.ndarray,
        dm0: np.ndarray = None,
    ) -> np.ndarray:
        """Generate the **skeleton** contribution of Hessian for current SCF component.

        Parameters
        ----------
        mo_coeff : np.ndarray
            Molecular orbital coefficients, shape ``[2, nao, nmo]``.
        mo_occ : np.ndarray
            Molecular orbital occupation numbers, shape ``[2, nmo]``.
        dm0 : np.ndarray, optional
            The **total** density matrix ``D^alpha + D^beta``, shape ``[nao, nao]``.

        Returns
        -------
        hess : np.ndarray
            The Hessian matrix for current SCF component, shape ``[natm, natm, 3, 3]``.
        """
        raise NotImplementedError

    @abstractmethod
    def generator_deriv1(self) -> callable:
        """Generate the function to compute the first-order derivative of core component.

        For UHF this is spin-independent (e.g. hcore), so the returned function is the
        same as in the restricted case: takes atom index, returns shape ``[3, nao, nao]``.

        If this component does not contribute (like nuclear repulsion), return None.
        """
        raise NotImplementedError


class UHessElecInteractAPI(ABC):
    """Abstract class for Hessian-related API for unrestricted SCF electronic interaction components.

    UHF Conventions
    ---------------

    Differences from the restricted version:
    - ``mo_coeff`` / ``mo_occ`` are passed as per-spin stacked arrays of shape
      ``[2, nao, nmo]`` and ``[2, nmo]`` respectively.
    - The first-order skeleton derivative in AO basis is **spin-resolved**:
      ``get_deriv1_ao`` returns ``[2, natm, 3, nao, nao]``.
      (UHF Fock matrix has different J/K coupling between alpha and beta channels, so
      the Fock skeleton derivative differs by spin even when the AO integrals don't.)
    - The half-transformation to bra uses spin-specific occupied orbitals, so
      ``get_deriv1_bra`` returns a *list* of two arrays whose last dimension differs
      (``[natm, 3, nao, nocc_alpha]`` and ``[natm, 3, nao, nocc_beta]``).
    - The response operator ``get_response_bra`` couples the two spin channels through
      the J part and decouples through the K part. It takes and returns a list of two
      arrays, of shape ``[..., nao, nocc_alpha]`` and ``[..., nao, nocc_beta]``.
    """

    @abstractmethod
    def make_skeleton_hess(
        self,
        mo_coeff: np.ndarray,
        mo_occ: np.ndarray,
    ) -> np.ndarray:
        """Generate the **skeleton** contribution of Hessian for current SCF component.

        Returns
        -------
        hess : np.ndarray
            The Hessian matrix for current SCF component, shape ``[natm, natm, 3, 3]``.
        """
        raise NotImplementedError

    @abstractmethod
    def get_deriv1_ao(self, mo_coeff: np.ndarray, mo_occ: np.ndarray) -> np.ndarray:
        """First order skeleton derivative in AO basis, spin-resolved.

        Returns
        -------
        deriv_ao : np.ndarray
            Shape ``[2, natm, 3, nao, nao]``.
        """
        raise NotImplementedError

    def get_deriv1_bra(self, mo_coeff: np.ndarray, mo_occ: np.ndarray) -> list[np.ndarray]:
        """First order skeleton derivative in half-transformed (bra) MO basis.

        Default implementation calls ``get_deriv1_ao`` and right-multiplies each spin block
        with the corresponding occupied coefficients.

        Returns
        -------
        deriv_bra : list[np.ndarray]
            A list ``[bra_alpha, bra_beta]`` where each entry has shape
            ``[natm, 3, nao, nocc_sigma]``.
        """
        deriv_ao = self.get_deriv1_ao(mo_coeff, mo_occ)
        out = []
        for s in range(2):
            occidx = mo_occ[s] > 1e-15
            mocc = mo_coeff[s][:, occidx]
            out.append(deriv_ao[s] @ mocc)
        return out

    @abstractmethod
    def make_response_preparation(self, mo_coeff: np.ndarray, mo_occ: np.ndarray):
        """Prepare the data for response calculation."""
        pass

    @abstractmethod
    def get_response_bra(self, bra: list[np.ndarray]) -> list[np.ndarray]:
        r"""Get the response contribution for current SCF component.

        Parameters
        ----------
        bra : list[np.ndarray]
            ``[bra_alpha, bra_beta]``. Each entry has shape ``[..., nao, nocc_sigma]``.
            Leading dimensions must agree across the two spins.

        Returns
        -------
        resp_bra : list[np.ndarray]
            ``[resp_alpha, resp_beta]`` with the same shapes as the inputs.
        """
        raise NotImplementedError

File: src/hessian/hess_trait_unrestricted.rs
use crate::prelude::*;

/// Abstract class for Hessian-related API for unrestricted SCF core components.
///
/// Difference to [`RHessCoreAPI`] is that we may need different signature. Basic ideas are exactly
/// the same.
pub trait UHessCoreAPI {
    /// Generate the **skeleton** contribution of Hessian for current SCF component.
    ///
    /// # Parameters
    ///
    /// - `mo_coeff` : shape `[nao, nmo_alpha]` and `[nao, nmo_beta]`. Molecular orbital
    ///   coefficients.
    /// - `mo_occ` : shape `[nmo_alpha]` and `[nmo_beta]`. Molecular orbital occupation numbers. In
    ///   usual cases, the occupied orbitals should have occupation 1 (unrestricted), and virtual
    ///   orbitals should have occupation 0.
    /// - `atm_list` : optional list of atom indices to compute the Hessian for. If `None`, all
    ///   atoms are computed.
    ///
    /// # Returns
    ///
    /// - `hess` : shape `[3, 3, natm, natm]`. The Hessian matrix for current SCF component, where
    ///   `natm = atm_list.len()` if `atm_list` is `Some`, else `mol.natm()`.
    ///
    ///   Note the hessian should be of indices `[s, t, B, A]` for column major.
    ///
    /// # See also
    ///
    /// [`RHessCoreAPI::make_skeleton_hess`]. Signature difference: `mo_coeff` and `mo_occ` type
    /// different.
    fn make_skeleton_hess(&mut self, mo_coeff: &[TsrView; 2], mo_occ: &[TsrView; 2], atm_list: Option<&[usize]>)
        -> Tsr;

    /// Generate the function to compute the first-order derivative of core component.
    ///
    /// This function works in atomic basis, so should have same implementation to restricted case.
    ///
    /// # Parameters (in closure)
    ///
    /// - `A` : usize. The atom index (global, in original molecule) for which the derivative is
    ///   taken.
    ///
    /// # Returns (in closure)
    ///
    /// - `deriv1` : shape `[nao, nao, 3]`. The first-order derivative of core component with
    ///   respect to the position of atom `A`.
    ///
    /// # See also
    ///
    /// [`RHessCoreAPI::generator_deriv1`]. Signature difference: no difference.
    fn generator_deriv1(&self) -> Box<dyn FnMut(usize) -> Tsr>;
}

/// Abstract class for Hessian-related API for restricted SCF electronic interaction components.
///
/// Difference to [`RHessElecInteractAPI`] is that we may need different signature. Basic ideas are
/// exactly the same.
pub trait UHessElecInteractAPI {
    /// Generate the **skeleton** contribution of Hessian for current SCF component.
    ///
    /// # Parameters
    ///
    /// - `mo_coeff` : shape `[nao, nmo_alpha]` and `[nao, nmo_beta]`. Molecular orbital
    ///   coefficients.
    /// - `mo_occ` : shape `[nmo_alpha]` and `[nmo_beta]`. Molecular orbital occupation numbers. In
    ///   usual cases, the occupied orbitals should have occupation 1, and virtual orbitals should
    ///   have occupation 0.
    /// - `atm_list` : optional list of atom indices to compute the Hessian for. If `None`, all
    ///   atoms are computed.
    ///
    /// # Returns
    ///
    /// - `hess` : shape `[3, 3, natm, natm]`. The Hessian matrix for current SCF component.
    ///
    ///   Note the hessian should be of indices `[s, t, B, A]` for column major.
    ///
    /// # See also
    ///
    /// [`RHessElecInteractAPI::make_skeleton_hess`]. Signature difference: `mo_coeff` and `mo_occ`
    /// type different.
    fn make_skeleton_hess(&mut self, mo_coeff: &[TsrView; 2], mo_occ: &[TsrView; 2], atm_list: Option<&[usize]>)
        -> Tsr;

    /// First order skeleton derivative in AO basis.
    ///
    /// # Parameters
    ///
    /// - `mo_coeff` : shape `[nao, nmo_alpha]` and `[nao, nmo_beta]`. Molecular orbital
    ///   coefficients.
    /// - `mo_occ` : shape `[nmo_alpha]` and `[nmo_beta]`. Molecular orbital occupation numbers.
    /// - `atm_list` : optional list of atom indices over which derivatives are computed.
    ///
    /// # Returns
    ///
    /// - `deriv_ao` : shape `[nao, nao, 3, natm, 2]`. The first-order skeleton derivative in AO
    ///   basis.
    ///
    /// # See also
    ///
    /// [`RHessElecInteractAPI::get_deriv1_ao`]. Signature difference: `mo_coeff` and `mo_occ` type
    /// different, output shape different.
    fn get_deriv1_ao(&mut self, mo_coeff: &[TsrView; 2], mo_occ: &[TsrView; 2], atm_list: Option<&[usize]>) -> Tsr;

    /// First order skeleton derivative in half-transformed MO basis.
    ///
    /// # Parameters
    ///
    /// - `mo_coeff` : shape `[nao, nmo_alpha]` and `[nao, nmo_beta]`. Molecular orbital
    ///   coefficients.
    /// - `mo_occ` : shape `[nmo_alpha]` and `[nmo_beta]`. Molecular orbital occupation numbers.
    /// - `atm_list` : optional list of atom indices over which derivatives are computed.
    ///
    /// # Returns
    ///
    /// - `deriv_bra` : shape `[nao, nocc_alpha, 3, natm]` and `[nao, nocc_beta, 3, natm]`. The
    ///   first-order skeleton derivative in half-transformed MO basis. Note that this function will
    ///   handle the order of occupied orbitals. If occupation number is not sorted contiguously,
    ///   you may be extra cautious to this function.
    ///
    /// # See also
    ///
    /// [`RHessElecInteractAPI::get_deriv1_bra`]. Signature difference: `mo_coeff` and `mo_occ` type
    /// different, output type different.
    fn get_deriv1_bra(
        &mut self,
        mo_coeff: &[TsrView; 2],
        mo_occ: &[TsrView; 2],
        atm_list: Option<&[usize]>,
    ) -> [Tsr; 2] {
        // Using `a` and `b` may be confusing for spin description. Using non-ASCII may be better.
        // To my suprise, using Han instead of Greek is actually better solution for better Unicode
        // programming. Well, the price to pay is a good monospaced font.
        // See also rust lint `mixed_script_confusables`, RFCS-2457, PEP 3131, UTS #39.
        // 我知道大家都喜欢看易语言的笑话；但实在是没想到用汉字真的是正解之一。
        // 本来觉得“升”“降”的标识度更高，但我们不是在搞二次量子化。
        // (正确说法是产生湮灭，但因为简谐振子解的原因习惯用升降算符了，大概算是我国文化特色)。
        // 感觉还是正常的“上”“下”更合适。
        let [上, 下] = [0, 1];

        let occidx = [mo_occ[上].view().greater(0).into_vec(), mo_occ[下].view().greater(0).into_vec()];
        let mocc = [mo_coeff[上].bool_select(-1, &occidx[上]), mo_coeff[下].bool_select(-1, &occidx[下])];
        let deriv1_ao = self.get_deriv1_ao(mo_coeff, mo_occ, atm_list);
        [&deriv1_ao % &mocc[上], &deriv1_ao % &mocc[下]]
    }

    /// Prepare the data for response calculation.
    ///
    /// # Parameters
    ///
    /// - `mo_coeff` : shape `[nao, nmo_alpha]` and `[nao, nmo_beta]`. Molecular orbital
    ///   coefficients.
    /// - `mo_occ` : shape `[nmo_alpha]` and `[nmo_beta]`. Molecular orbital occupation numbers.
    ///
    /// # See also
    ///
    /// [`RHessElecInteractAPI::make_response_preparation`]. Signature difference: `mo_coeff` and
    /// `mo_occ` type different.
    fn make_response_preparation(&mut self, mo_coeff: &[TsrView; 2], mo_occ: &[TsrView; 2]);

    /// Get the response contribution for current SCF component.
    ///
    ///
    /// # Parameters
    ///
    /// - `bra` : shape `[nao, nocc_alpha, ...]` and `[nao, nocc_beta, ...]`. The bra part for
    ///   response calculation.
    ///
    /// # Returns
    ///
    /// - `resp_bra` : shape `[nao, nocc_alpha, ...]` and `[nao, nocc_beta, ...]`. The response
    ///   potential (related to second order of density matrix derivative to energy).
    ///
    /// # See also
    ///
    /// [`RHessElecInteractAPI::get_response_bra`]. Signature difference: input and output type
    /// different.
    fn get_response_bra(&self, bra: &[TsrView; 2]) -> [Tsr; 2];
}
